### PR TITLE
FINERACT-1684: Business date - part 2

### PR DESCRIPTION
**Pull request by adamsaghy**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/adhocquery/service/AdHocScheduledJobRunnerServiceImpl.java [hunk 1]
```diff
@@ -42,12 +42,14 @@ public class AdHocScheduledJobRunnerServiceImpl implements AdHocScheduledJobRunn
     private static final Logger LOG = LoggerFactory.getLogger(AdHocScheduledJobRunnerServiceImpl.class);
     private final AdHocReadPlatformService adHocReadPlatformService;
     private final JdbcTemplate jdbcTemplate;
+    private final DatabaseSpecificSQLGenerator sqlGenerator;
 
     @Autowired
-    public AdHocScheduledJobRunnerServiceImpl(final JdbcTemplate jdbcTemplate, final AdHocReadPlatformService adHocReadPlatformService) {
+    public AdHocScheduledJobRunnerServiceImpl(final JdbcTemplate jdbcTemplate, final AdHocReadPlatformService adHocReadPlatformService,
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/adhocquery/service/AdHocScheduledJobRunnerServiceImpl.java]: Could be replaced with `@RequiredArgsConstructor`

File: fineract-provider/src/main/java/org/apache/fineract/adhocquery/service/AdHocScheduledJobRunnerServiceImpl.java [hunk 2]
```diff
@@ -42,12 +42,14 @@ public class AdHocScheduledJobRunnerServiceImpl implements AdHocScheduledJobRunn
     private static final Logger LOG = LoggerFactory.getLogger(AdHocScheduledJobRunnerServiceImpl.class);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/adhocquery/service/AdHocScheduledJobRunnerServiceImpl.java]: I know you haven't introduced it but if you're touching the constructor already, we could replace this with an `@Slf4J`

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java [hunk 3]
```diff
@@ -26,29 +31,25 @@
...
-
     public static final String CONTEXT_TENANTS = "tenants";
-
     private static final ThreadLocal<String> contextHolder = new ThreadLocal<>();
-
-    private static final ThreadLocal<FineractPlatformTenant> tenantcontext = new ThreadLocal<>();
-
+    private static final ThreadLocal<FineractPlatformTenant> tenantContext = new ThreadLocal<>();
     private static final ThreadLocal<String> authTokenContext = new ThreadLocal<>();
+    private static final ThreadLocal<HashMap<BusinessDateType, LocalDate>> businessDateContext = new ThreadLocal<>();
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java:38]: Do you think we could switch the `HashMap` into a regular `Map` interface? I'd say let's try to be as independent from a specific implementation as possible.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java [hunk 4]
```diff
@@ -63,12 +64,55 @@ public static void clearDataSourceContext() {
...
+    public static ActionContext getActionContext() {
+        return actionContext.get() == null ? ActionContext.DEFAULT : actionContext.get();
+    }
+
+    public static void setActionContext(ActionContext context) {
+        Assert.notNull(context, "context cannot be null");
+        actionContext.set(context);
+    }
+
+    public static FineractContext syncDown() {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java]: `syncDown` is a bit confusing for me cause it's not "syncing" anything per se. I'd simply call it something like `ThreadLocalContextUtil#getContext`/`ThreadLocalContextUtil#toContext`

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java [hunk 5]
```diff
@@ -63,12 +64,55 @@ public static void clearDataSourceContext() {
...
+        return new FineractContext(getDataSourceContext(), getTenant(), getAuthToken(), getBusinessDates(), getActionContext());
+    }
+
+    public static void syncUp(final FineractContext fineractContext) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java]: Same as above. Probably fromContext would be a better naming.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java [hunk 6]
```diff
@@ -63,12 +64,55 @@ public static void clearDataSourceContext() {
...
+        return authTokenContext.get();
+    }
+
     public static void setAuthToken(final String authToken) {
         authTokenContext.set(authToken);
     }
 
-    public static String getAuthToken() {
-        return authTokenContext.get();
+    public static HashMap<BusinessDateType, LocalDate> getBusinessDates() {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java:76]: No `HashMap` here either, simple `Map` is fine.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/domain/FineractContext.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.fineract.infrastructure.businessdate.domain.BusinessDateType;
+
+@AllArgsConstructor
+@Getter
+public class FineractContext implements Serializable {
+
+    private final String contextHolder;
+    private final FineractPlatformTenant tenantContext;
+    private final String authTokenContext;
+    private final HashMap<BusinessDateType, LocalDate> businessDateContext;
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/domain/FineractContext.java:35]: Simple Map here too?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/SchedulerJobListener.java [hunk 8]
```diff
@@ -39,25 +43,23 @@
...
-
     private final AppUserRepositoryWrapper userRepository;
-
     private final GrantedAuthoritiesMapper authoritiesMapper = new NullAuthoritiesMapper();
+    private final BusinessDateReadPlatformService businessDateReadPlatformService;
+    private int stackTraceLevel = 0;
 
     @Autowired
-    public SchedulerJobListener(final SchedularWritePlatformService schedularService, final AppUserRepositoryWrapper userRepository) {
+    public SchedulerJobListener(final SchedularWritePlatformService schedularService, final AppUserRepositoryWrapper userRepository,
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/SchedulerJobListener.java]: `@RequiredArgsConstructor`?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/SchedulerTriggerListener.java [hunk 9]
```diff
@@ -42,11 +46,14 @@ public class SchedulerTriggerListener implements TriggerListener {
     private final SchedularWritePlatformService schedularService;
     private final TenantDetailsService tenantDetailsService;
 
+    private final BusinessDateReadPlatformService businessDateReadPlatformService;
+
     @Autowired
-    public SchedulerTriggerListener(final SchedularWritePlatformService schedularService, final TenantDetailsService tenantDetailsService) {
+    public SchedulerTriggerListener(final SchedularWritePlatformService schedularService, final TenantDetailsService tenantDetailsService,
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/service/SchedulerTriggerListener.java]: `@RequiredArgsConstructor`?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/TenantAwareTenantIdentifierFilter.java [hunk 10]
```diff
@@ -66,18 +70,21 @@ public class TenantAwareTenantIdentifierFilter extends GenericFilterBean {
...
+
     private final String tenantRequestHeader = "Fineract-Platform-TenantId";
     private final boolean exceptionIfHeaderMissing = true;
     private final String apiUri = "/api/v1/";
 
     @Autowired
     public TenantAwareTenantIdentifierFilter(final BasicAuthTenantDetailsService basicAuthTenantDetailsService,
             final ToApiJsonSerializer<PlatformRequestLog> toApiJsonSerializer, final ConfigurationDomainService configurationDomainService,
-            final CacheWritePlatformService cacheWritePlatformService) {
+            final CacheWritePlatformService cacheWritePlatformService, BusinessDateReadPlatformService businessDateReadPlatformService) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/TenantAwareTenantIdentifierFilter.java]: `@RequiredArgsConstructor`?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/sms/scheduler/SmsMessageScheduledJobServiceImpl.java [hunk 11]
```diff
@@ -65,14 +65,14 @@
 @Service
 public class SmsMessageScheduledJobServiceImpl implements SmsMessageScheduledJobService {
 
+    private static final Logger LOG = LoggerFactory.getLogger(SmsMessageScheduledJobServiceImpl.class);
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/sms/scheduler/SmsMessageScheduledJobServiceImpl.java]: `@Slf4j`?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/account/service/PortfolioAccountReadPlatformServiceImpl.java [hunk 12]
```diff
@@ -294,18 +294,22 @@ public PortfolioAccountData mapRow(final ResultSet rs, @SuppressWarnings("unused
 
     private static final class PortfolioLoanAccountRefundByTransferMapper implements RowMapper<PortfolioAccountData> {
 
-        private final String schemaSql;
+        private String schemaSql;
+        private final DatabaseSpecificSQLGenerator sqlGenerator;
 
         PortfolioLoanAccountRefundByTransferMapper(DatabaseSpecificSQLGenerator sqlGenerator) {
+            this.sqlGenerator = sqlGenerator;
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/account/service/PortfolioAccountReadPlatformServiceImpl.java:301]: `@RequiredArgsConstructor`?

**Review by galovics — CHANGES_REQUESTED** [review]
Couple minor things but overall really nice work.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java (hunk 3 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/ThreadLocalContextUtil.java:38]: Map is not supporting serialization.

**Review by galovics — APPROVED** [review]

**Pull request merged**